### ISSUE TITLE
Odd wasm behaviour when evaluating input equality

### ISSUE THREAD

**Issue by nickhiggs** [open-policy-agent/opa#3128]
## Expected Behavior

I was experimenting with Wasm and was attempting to translate an existing rego policy which included:

```
allow["open"] {
  any([input.path == "/swagger.json", input.path == "/swagger.yaml"])
}
```

which works well with normal evaluation.

## Actual Behavior

The wasm compiled & evaluated policy doesn't evaluate the rule to true when input is `{"path":"/swagger.json"}`.

## Steps to Reproduce the Problem

Apply this patch taken from current master (2b2d73eb):
```
diff --git a/internal/wasm/sdk/opa/opa_test.go b/internal/wasm/sdk/opa/opa_test.go
index 5638a41e..7421e434 100644
--- a/internal/wasm/sdk/opa/opa_test.go
+++ b/internal/wasm/sdk/opa/opa_test.go
@@ -57,6 +57,16 @@ func TestOPA(t *testing.T) {
 			},
 			WantErr: "",
 		},
+		{
+			Description: "Eval equality",
+			Policy:      `a = input == true`,
+			Query:       "data.p.a = x",
+			Evals: []Eval{
+				Eval{Input: `false`, Result: `{{"x": false}}`},
+				Eval{Input: `true`, Result: `{{"x": true}}`},
+			},
+			WantErr: "",
+		},
 		{
 			Description: "Only data changing",
 			Policy:      `a = data.q`,
```

which fails when I'd expect it to succeed.

**Comment by srenatus** [comment]
Thanks for the detailed report. I'll have a closer look tomorrow, from the first impression, it looks wrong indeed.

**Comment by srenatus** [comment]
Those test cases can by misleading, since they expect their policy rego in the way the compiler would rewrite it to. With #3090, we can use `opa eval -t wasm` directly to examine the Wasm behavior:

```
scratch/issue-3128 % ../../opa eval -t rego --format=pretty -d p.rego -i input.json data.p.allow
true
scratch/issue-3128 % ../../opa eval -t wasm --format=pretty -d p.rego -i input.json data.p.allow
true
scratch/issue-3128 %
```

where `p.rego` is 
```rego
package p

allow {
  input == true
}
```
and `input.json` is only `true`.

However, querying directly, we find
```
scratch/issue-3128 % ../../opa eval -t rego --format=pretty -i input.json 'x = input == true'
+------+
|  x   |
+------+
| true |
+------+
scratch/issue-3128 % ../../opa eval -t wasm --format=pretty -i input.json 'x = input == true'
1 error occurred: expected exactly one term but got:
scratch/issue-3128 %
```
Which does point at some issue since it's not the same: there's an error for wasm.

**Comment by srenatus** [comment]
BTW if you look at what the compiler rewrites that test case to, you'll find
```rego
package p
a = true { input = true}
```
With that snippet the added test still fails, but for understandable reasons: if `input` is false, the `input = true` unification fails, and the result becomes `undefined` (not `false`).

**Comment by srenatus** [comment]
Besides the issue above, I'm having trouble reproducing your problem using Wasm bundles. Please have a look:

```
scratch/issue-3128 % cat i.rego
package i

allow["open"] {
  any([input.path == "/swagger.json", input.path == "/swagger.yaml"])
}
scratch/issue-3128 % cat ii.json
{ "path": "/swagger.json" }
scratch/issue-3128 % opa build -t wasm i.rego -e data.i.allow
scratch/issue-3128 % opa eval -b bundle.tar.gz data.i.allow -i ii.json
{
  "result": [
    {
      "expressions": [
        {
          "value": [
            "open"
          ],
          "text": "data.i.allow",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
scratch/issue-3128 %
```

So, I guess you've picked a different path. What is it? 🕵️

**Comment by tsandall** [comment]
@ashutosh-narkar can you take a look at the discrepancy that @srenatus pointed out in the [comment](https://github.com/open-policy-agent/opa/issues/3128#issuecomment-774984367) above?

**Comment by srenatus** [comment]
This has been resolved recently; I think with #3271  👇 
```
# opa eval -t rego --format=pretty -i input.json 'x = input == true'
+------+
|  x   |
+------+
| true |
+------+
# opa eval -t wasm --format=pretty -i input.json 'x = input == true'
+------+
|  x   |
+------+
| true |
+------+
```
